internal-packages/workflow-designer-ui/src/editor/properties-panel/file-node-properties-panel/use-web-page-file-node.ts

```diff
@@ -1,70 +1,87 @@
 import type { FileData, FileNode } from "@giselle-sdk/data-type";
-import { useState } from "react";
+import { useWorkflowDesigner } from "giselle-sdk/react";
+import { useCallback, useState } from "react";
 
-export function useWebPageFileNode(_node: FileNode) {
+export function useWebPageFileNode(node: FileNode) {
+	const { isLoading, updateNodeDataContent, uploadFile, fetchWebPageFiles } =
+		useWorkflowDesigner();
 	const [urls, setUrls] = useState("");
 	const [format, setFormat] = useState<"html" | "markdown">("markdown");
-	const [isFetching, setIsFetching] = useState(false);
-	const [fetchStatuses, setFetchStatuses] = useState<
-		Record<string, "fetching" | "success" | "error">
-	>({});
-	const [files, setFiles] = useState<FileData[]>([]);
 	const [urlError, setUrlError] = useState("");
 
-	const onFetch = async () => {
+	// onFetch now uploads web page content as files using uploadFile, and updates webPages metadata for display
+	const onFetch = useCallback(async () => {
 		const urlList = urls
 			.split("\n")
 			.map((u) => u.trim())
-			.filter(Boolean);
+			.filter((u: string) => Boolean(u));
 		if (urlList.length === 0) return;
-		setIsFetching(true);
 		setUrlError("");
-		setFetchStatuses(
-			Object.fromEntries(urlList.map((url) => [url, "fetching"])),
-		);
-		// Simulate async fetch for each URL
-		await Promise.all(
-			urlList.map(
-				(url, i) =>
-					new Promise<void>((resolve) => {
-						setTimeout(
-							() => {
-								setFetchStatuses((prev) => ({ ...prev, [url]: "success" }));
-								setFiles((prev) => [
-									...prev,
-									{
-										id: `fl-${url.replace(/[^a-zA-Z0-9]/g, "")}-${Date.now()}`,
-										name: url,
-										type: format === "markdown" ? "text/markdown" : "text/html",
-										size: 0,
-										status: "uploaded",
-										uploadedAt: Date.now(),
-									},
-								]);
-								resolve();
-							},
-							800 + i * 400,
-						); // staggered for effect
-					}),
-			),
-		);
-		setIsFetching(false);
-	};
 
-	const onRemoveFile = (file: FileData) => {
-		setFiles((prev) => prev.filter((f) => f.id !== file.id));
-	};
+		try {
+			const webPageFileResults = await fetchWebPageFiles({
+				urls: urlList,
+				format,
+			});
+
+			// Convert content string to File for browser upload
+			const filesToUpload = webPageFileResults.map((r) => {
+				const blob = new Blob([r.content], { type: r.mimeType });
+				return new File([blob], r.fileName, { type: r.mimeType });
+			});
+			const fileDataList = webPageFileResults.map((r) => r.fileData);
+
+			if (filesToUpload.length > 0) {
+				await uploadFile(filesToUpload, node, {
+					onError: (errorMessage) => {
+						setUrlError(errorMessage);
+					},
+				});
+			}
+
+			// Update node content for all file statuses
+			updateNodeDataContent(node, {
+				files: fileDataList,
+			});
```

Comment:
I think it is implemented with the following steps:
1. With fetchWebPageFiles, fetch the input URL in the configured format
2. Convert the retrieved data into files and upload them

I think it's a good implementation that reuses existing implementation, but I believe there's waste in the processing, so I think it would be better to have "fetch URL in format and save" as a single function, loop through it for each URL, and update the status accordingly (this doesn't need to be in this Pull request. It's probably complicated in various ways, so I think I'll try implementing it myself).

Reply:
Thank you for your polite comments!

You are absolutely right. I recognize that the overhead is significant, but because implementation seemed like it would take a lot of time, I proceeded with an approach that leverages the existing file update mechanism.

I would be grateful if you could re-architect it when you have a moment!